### PR TITLE
ds-querier: parse datasourceUid correctly

### PR DESCRIPTION
**Pull request by asimpson**

### CODE REVIEW

File: pkg/registry/apis/query/testdata/legacy-ds-uid.json
Comment (samjewell) [inline pkg/registry/apis/query/testdata/legacy-ds-uid.json:1]: I searched the folder `/pkg/registry/apis/query` for `legacy-ds-uid` and found nothing. Then searched for `testdata` and got a LOT of hits: 29 results total.

Outside the scope of this PR of course, but I'm wondering whether we could rename this folder to give it a more unique name 🤔 


In any case, perhaps more in scope now for this PR, I'm wondering why we have this test-case in a JSON file, and not as a more typical entry in a parameterised Go testing struct or something. Any thoughts on this? It feels quite surprising to me, and not idiomatic 🤔

File: pkg/registry/apis/query/testdata/legacy-ds-uid.json [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        "datasource": {
+          "type": "__expr__",
+          "uid": "__expr__"
+        },
+        "expression": "$B"
+      },
+      {
+        "refId": "A",
+        "queryType": "",
+        "datasourceUid": "XYX",
```
Comment (samjewell) [inline pkg/registry/apis/query/testdata/legacy-ds-uid.json:23]: `XYX`? Could you put something slightly more meaningful in here? Dunno, maybe something like:

```suggestion
        "datasourceUid": "test_case_datasource_UID",
```

Or even:
```suggestion
        "datasourceUid": "foo",
```

File: pkg/registry/apis/query/testdata/legacy-ds-uid.json [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          "uid": "__expr__"
+        },
+        "expression": "$A + 10"
+      }
+    ]
+  },
+  "expect": {
+    "requests": [
+      {
+        "uid": "XYX",
```
Comment (samjewell) [inline pkg/registry/apis/query/testdata/legacy-ds-uid.json:44]:
```suggestion
        "uid": "test_case_datasource_UID",
```

File: pkg/registry/apis/query/testdata/legacy-ds-uid.json [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+          "from": "now-6",
+          "to": "now",
+          "queries": [
+            {
+              "refId": "A",
+              "datasource": {
+                "uid": "${datasource}",
+                "type": ""
+              },
+              "datasourceUid": "XYX"
```
Comment (samjewell) [inline pkg/registry/apis/query/testdata/legacy-ds-uid.json:56]:
```suggestion
              "datasourceUid": "test_case_datasource_UID"
```

File: pkg/registry/apis/query/testdata/legacy-ds-uid.json [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    ],
+    "expressions": [
+      {
+        "id": 2,
+        "refId": "B",
+        "type": "math",
+        "properties": {
+          "expression": "$A + 10"
+        }
+      },
```
Comment (samjewell) [inline pkg/registry/apis/query/testdata/legacy-ds-uid.json:70]: I like test cases to be as small and simple as possible. To me it seems this block is outside the scope of the new code we're adding here, and the new specification we're trying to set for this code. Can we remove this block completely? Can we shorten this test in any other ways?

File: pkg/registry/apis/query/parser.go [hunk 5]
```diff
@@ -80,7 +80,7 @@ func (p *queryParser) parseRequest(ctx context.Context, input *query.QueryDataRe
 			return rsp, MakePublicQueryError(q.RefID, "multiple queries with same refId")
 		}
 
-		ds, err := p.getValidDataSourceRef(ctx, q.Datasource, q.DatasourceID)
+		ds, err := p.getValidDataSourceRef(ctx, q)
```
Comment (samjewell) [inline pkg/registry/apis/query/parser.go:83]: I like this refactoring. It feels like we're removing the "Data clump" code smell by making this refactoring.
But I would have liked to see the refactoring as its own separate commit - I think it would have made this PR easier to review.

File: pkg/registry/apis/query/parser.go [hunk 6]
```diff
@@ -194,16 +194,30 @@ func getTimeRangeForQuery(parentTimerange, queryTimerange *data.TimeRange) data.
...
+			return nil, NewErrorWithRefID(dataQuery.RefID, fmt.Errorf("missing datasource reference or id"))
 		}
+
 		if p.legacy == nil {
 			return nil, fmt.Errorf("legacy datasource lookup unsupported (id:%d)", id)
 		}
 		return p.legacy.GetDataSourceFromDeprecatedFields(ctx, "", id)
 	}
+
+	if ds.UID[0] == '$' {
```
Comment (samjewell) [inline pkg/registry/apis/query/parser.go:212]: Why are you only checking the first character here? Isn't this too loose?
Can we check the full string exactly matches instead? 

```suggestion
	if ds.UID == '${datasource}' {
```

I feel like this is safer somehow, but I don't have much context here.
Where is the equivalent, old version of the code, so we can compare what was done when this "Permanent migration" (as Kyle called it) was first introduced?

Also, is the `Type == ""` in this case too? Should this code be moved under the `if ds.Type == "" {` check on line 221? Again I think it's worth checking what the logic was on the original code-path, when this migration was introduced.

**Review by samjewell — APPROVED** [review]
Thanks for this work - great to see this moving forward.

I hesitated to approve, but in the end I wanted to make sure you weren't blocked. But I don't have huge confidence in this. I'd like to compare the specification (ie the test suite) for the old code-path with the new, and gain confidence than they match exactly. And make the same check for the code-behaviour too I guess (given that the code might not be fully tested).

Otherwise I feel like we (or our customers) may just run into more differences in future.

File: pkg/registry/apis/query/parser.go (hunk 6 above)
Comment (asimpson) [reply pkg/registry/apis/query/parser.go:212]: I've seen two values `$datasource` and `${datasource}`. So I went with `$` in this case. :man_shrugging: 

I put the check after the `ds == nil` check and before the `Type == ""` check since the `Type` check path needs a `uid` which we won't have if do this check last.

File: pkg/registry/apis/query/parser.go (hunk 5 above)
Comment (asimpson) [reply pkg/registry/apis/query/parser.go:83]: ah my bad, I needed to pass the entire `DataQuery` struct to call `.Get("datasourceUid")` later so I kept it together so the commit could "work on its own".

File: pkg/registry/apis/query/testdata/legacy-ds-uid.json (hunk 4 above)
Comment (asimpson) [reply pkg/registry/apis/query/testdata/legacy-ds-uid.json:70]: Ah yeah, good call. I was _very_ focused on this case working with expressions that I forgot to clear out the test :trophy:

File: pkg/registry/apis/query/testdata/legacy-ds-uid.json
Comment (asimpson) [reply pkg/registry/apis/query/testdata/legacy-ds-uid.json:1]: I couldn't find any previous test cases for this behavior.  It seems that it's never been a problem because alerting has [always pulled](https://github.com/grafana/grafana/blob/f8a581357373f4911f411d2ef3833ccaf041d5e5/pkg/services/ngalert/eval/eval.go#L362) the datasource `uid` from `datasourceUid` like this PR does. 

The test case is in a JSON file because the existing tests are organized that way, the structure of these files is:

```json
{
"description": "A brief description of the test case",
"input": "The query object we're testing",
"expect": "The shape of the JSON we expect"
}
```

This seems to work well and therefore I didn't think we needed to change it. WDYT?

Comment (samjewell) [reply pkg/registry/apis/query/testdata/legacy-ds-uid.json:1]: > I couldn't find any previous test cases for this behavior

💔 

> It seems that it's never been a problem because alerting has [always pulled](https://github.com/grafana/grafana/blob/f8a581357373f4911f411d2ef3833ccaf041d5e5/pkg/services/ngalert/eval/eval.go#L362) the datasource uid from datasourceUid like this PR does.

Oh weird - ok thanks a lot for the link. Does that mean there are two code paths we can remove when all our work here is complete, and not just one? Ie. both the alerting one you've linked to here, and the one Ryan wants to remove soon?

> WDYT?

👍 yep, thanks for the explainer - I'm fine to keep it for now, sure thing :-)

Comment (asimpson) [reply pkg/registry/apis/query/testdata/legacy-ds-uid.json:1]: > Does that mean there are two code paths we can remove when all our work here is complete, and not just one? Ie. both the alerting one you've linked to here, and the one Ryan wants to remove soon?

Possibly, we'll have to evaluate a lot of things once we're fully deployed.
